### PR TITLE
Reimplement date field validator as validation rule

### PR DESCRIPTION
**Pull request by 1ec5**
Replaced the date format validator from #81 with a user-friendlier check within the existing invalid format validation rule. The new rule comes with a one-click option to reformat a malformed but understandable date, wrapping around months and years as necessary, as well as an option to remove the malformed date.

 July 4, 1776  

Fixes OpenHistoricalMap/issues#445.

### CODE REVIEW

File: test/spec/util/util.js [hunk 1]
```diff
@@ -225,4 +225,32 @@ describe('iD.util', function() {
...
+    describe('utilNormalizeDateString', function() {
+        it('pads dates', function() {
+            expect(iD.utilNormalizeDateString('1970-01-01').value).to.eql('1970-01-01');
+            expect(iD.utilNormalizeDateString('1970-01-1').value).to.eql('1970-01-01');
+            expect(iD.utilNormalizeDateString('1970-1-01').value).to.eql('1970-01-01');
+            expect(iD.utilNormalizeDateString('-1-1-01').value).to.eql('-0001-01-01');
+            expect(iD.utilNormalizeDateString('123').value).to.eql('0123');
+            expect(iD.utilNormalizeDateString('-4003').value).to.eql('-4003'); // beyond displayable year range but still valid
+            expect(iD.utilNormalizeDateString('31337').value).to.eql('31337');
+            expect(iD.utilNormalizeDateString('-31337').value).to.eql('-31337');
```
Comment (1ec5) [inline test/spec/util/util.js:318]: These years are well beyond the minimum year of 3999&nbsp;BCE set in OpenHistoricalMap/issues#28 and the maximum year equivalent to the current year at any given time. However, these year thresholds are only for display and filtering purposes. A truly primordial natural feature shouldn’t prevent the user from saving, nor should a building slated for demolition in a year’s time. If we want these to be hard limits, then the API should enforce them and expose them via the `capabilities` API method.

File: test/spec/util/util.js [hunk 2]
```diff
@@ -225,4 +225,32 @@ describe('iD.util', function() {
...
+        it('wraps dates', function() {
+            expect(iD.utilNormalizeDateString('1970-01-00').value).to.eql('1969-12-31');
+            expect(iD.utilNormalizeDateString('1970-12-32').value).to.eql('1971-01-01');
+            expect(iD.utilNormalizeDateString('1970-02-29').value).to.eql('1970-03-01');
+            expect(iD.utilNormalizeDateString('1970-00').value).to.eql('1969-12');
+            expect(iD.utilNormalizeDateString('1970-23').value).to.eql('1971-11'); // no EDTF for now
+        });
+        it('rejects malformed dates', function() {
+            expect(iD.utilNormalizeDateString('1970-01--1')).to.eql(null);
+            expect(iD.utilNormalizeDateString('197X')).to.eql(null); // no EDTF for now
```
Comment (1ec5) [inline test/spec/util/util.js:329]: Validating the `start_date:edtf` and `end_date:edtf` keys should wait until we have dedicated fields for these keys. Parsing EDTF is somewhat more complex than parsing ordinary ISO&nbsp;8601, since we can’t lean on the browser to bear the brunt of parsing them.

File: modules/util/util.js [hunk 3]
```diff
@@ -542,3 +542,65 @@ export function utilUnicodeCharsCount(str) {
...
     return Array.from(str).slice(0, limit).join('');
 }
+
+// Simple polyfill for String.prototype.padStart().
+function padStart(str, width, pad) {
+    var padded = Array(width);
+    padded.fill(pad);
+    padded = padded.join('');
+    return padded.substr(0, width - str.length) + str;
+}
```
Comment (1ec5) [inline modules/util/util.js]: PhantomJS throws an error when using `String.prototype.padStart()`, probably because this fork of iD is based on an older mainline iD from before it dropped support for IE for openstreetmap#8811.

File: modules/util/util.js [hunk 4]
```diff
@@ -542,3 +542,65 @@ export function utilUnicodeCharsCount(str) {
...
+    if (match !== null) {
+        // Manually construct a Date.
+        // Passing the string directly into the Date constructor would throw an error on negative years.
+        date = new Date(0);
+        date.setUTCFullYear(parseInt((match[1] || '') + match[2], 10));
+        if (match[3]) date.setUTCMonth(parseInt(match[3], 10) - 1); // 0-based
+        if (match[4]) date.setUTCDate(parseInt(match[4], 10));
+    } else {
+        // Fall back on whatever the browser can parse into a date.
+        date = new Date(raw);
```
Comment (1ec5) [inline modules/util/util.js:651]: This fallback is what allows for some natural language syntax to be parsed successfully. The specific supported natural language formats depends on the browser, language settings, etc., so it’s only a best-effort fallback. Anything produced by this code path still results in an error presented to the user, but there’s a suggestion to reformat the tag value to what the `Date` constructor interprets the value as.

File: test/spec/util/util.js [hunk 5]
```diff
@@ -225,4 +225,32 @@ describe('iD.util', function() {
...
+            expect(iD.utilNormalizeDateString('-4003').value).to.eql('-4003'); // beyond displayable year range but still valid
+            expect(iD.utilNormalizeDateString('31337').value).to.eql('31337');
+            expect(iD.utilNormalizeDateString('-31337').value).to.eql('-31337');
+        });
+        it('wraps dates', function() {
+            expect(iD.utilNormalizeDateString('1970-01-00').value).to.eql('1969-12-31');
+            expect(iD.utilNormalizeDateString('1970-12-32').value).to.eql('1971-01-01');
+            expect(iD.utilNormalizeDateString('1970-02-29').value).to.eql('1970-03-01');
+            expect(iD.utilNormalizeDateString('1970-00').value).to.eql('1969-12');
+            expect(iD.utilNormalizeDateString('1970-23').value).to.eql('1971-11'); // no EDTF for now
```
Comment (1ec5) [inline test/spec/util/util.js:325]: I’ve already made the mistake of [entering EDTF year-season combinations](https://www.openhistoricalmap.org/changeset/44781) into `start_date` and `end_date` a few times. It’ll be nice to get warned about this mistake instead of having to notice that the time slider never filters the feature out.

File: modules/ui/sections/map_daterange.js [hunk 6]
```diff
@@ -30,6 +27,9 @@ export function uiSectionDateRange(context) {
     // see modules/renderer/features.js checkDateFilter() which applies the filters
     // see modules/renderer/features.js update() which updates URL params
 
+    const MIN_YEAR = context.minYear();
+    const MAX_YEAR = context.maxYear();
```
Comment (1ec5) [inline modules/ui/sections/map_daterange.js]: This change is unrelated to the validator rule. I had originally planned to make the validator enforce this minimum and maximum year on the `start_date` and `end_date` tags before realizing that the range is primarily for display purposes.

File: modules/util/util.js (hunk 3 above)
Comment (1ec5) [reply modules/util/util.js]: After rebasing onto the latest staging branch, tests pass even when using the built-in `String.prototype.padStart()`.

File: modules/ui/sections/map_daterange.js (hunk 6 above)
Comment (1ec5) [reply modules/ui/sections/map_daterange.js]: I ended up removing this change from the PR, since it’s based on the date filter feature that hasn’t been cherry-picked yet.

**Comment by 1ec5** [comment]
I rebased the most substantial part of this PR onto the latest staging branch for OpenHistoricalMap/issues#395. I omitted the commits that rolled back #81, which have been completely superseded by this PR.

**Comment by batpad** [comment]
@1ec5 this looks fantastic, thank you so much!

@danrademacher am a bit confused with the branches :( - but do you guys have everything you need to integrate? Let us know if we can help!

**Comment by danrademacher** [comment]
@batpad Yes, I think we have what we need -- @1ec5 helped chart the way to a new `staging` that's based on latest upstream, which is what we have now.

**Comment by danrademacher** [comment]
Just tested this locally -- the screenshots should have alerted me to how awesome this is, but using it live re-inforced -- it is very cool!

**Comment by todrobbins** [comment]
@danrademacher is this now live on https://www.openhistoricalmap.org/?

**Comment by danrademacher** [comment]
Unfortunately, no. We have been having a bunch of problems with our local development setup after catching up with upstream OSM and iD, and iD is currently failing on our staging site -- but for reasons that we're still trying to figure out... Documented in https://github.com/OpenHistoricalMap/ohm-deploy/pull/189 (.local dev issues) and at https://github.com/OpenHistoricalMap/issues/issues/476 (iD failing on staging). We escalated this to get more help from our friends at DevSeed and I hope tomorrow we will have progress on local dev that will allow us to fix new iD inside rails and get this out to the community!
